File: tier1_foundation/two_pointers/problems/p522_longest_uncommon_subsequence_ii.c
/*
 * P522: Longest Uncommon Subsequence II [PREMIUM] (Medium)
 * https://leetcode.com/problems/longest-uncommon-subsequence-ii/
 * Topics: Array, Hash Table, Two Pointers, String, Sorting
 *
 * Given an array of strings strs, return the length of the longest uncommon subsequence between them. If the longest uncommon subsequence does not exist, return -1.
 * An uncommon subsequence between an array of strings is a string that is a subsequence of one string but not the others.
 * A subsequence of a string s is a string that can be obtained after deleting any number of characters from s.
 * Example 1:
 *     Input: strs = ["aba","cdc","eae"]
 *     Output: 3
 *
 * Example 2:
 *     Input: strs = ["aaa","aaa","aa"]
 *     Output: -1
 *
 * Constraints:
 *     - 2 <= strs.length <= 50
 *     - 1 <= strs[i].length <= 10
 *     - strs[i] consists of lowercase English letters.
 *
 * Template (python3):
 *     class Solution:
 *         def findLUSlength(self, strs: List[str]) -> int:
 */

#include "io.h"
#include <stdlib.h>
#include <string.h>

static int is_subseq(const char *a, const char *b)
{
    abort();
}

static int cmp_len_desc(const void *a, const void *b)
{
    abort();
}

static int findLUSlength(const char **strs, int n)
{
    abort();
}

int main(void)
{
    int n_line_count;
    int *n_arr = read_ints(&n_line_count);
    int n = n_arr[0];
    free(n_arr);
    const char **strs = malloc(n * sizeof(const char *));
    for (int i = 0; i < n; i++) {
        strs[i] = read_line();
    }
    write_int(findLUSlength(strs, n));
    for (int i = 0; i < n; i++) free((void *)strs[i]);
    free(strs);
    return 0;
}

File: tier1_foundation/two_pointers/problems/p522_longest_uncommon_subsequence_ii.mjs
/**
 * P522: Longest Uncommon Subsequence II [PREMIUM] (Medium)
 * https://leetcode.com/problems/longest-uncommon-subsequence-ii/
 * Topics: Array, Hash Table, Two Pointers, String, Sorting
 * 
 * Given an array of strings strs, return the length of the longest uncommon subsequence between them. If the longest uncommon subsequence does not exist, return -1.
 * An uncommon subsequence between an array of strings is a string that is a subsequence of one string but not the others.
 * A subsequence of a string s is a string that can be obtained after deleting any number of characters from s.
 * Example 1:
 *     Input: strs = ["aba","cdc","eae"]
 *     Output: 3
 * 
 * Example 2:
 *     Input: strs = ["aaa","aaa","aa"]
 *     Output: -1
 * 
 * Constraints:
 *     - 2 <= strs.length <= 50
 *     - 1 <= strs[i].length <= 10
 *     - strs[i] consists of lowercase English letters.
 */

function solve(strs) {
  throw new Error("NotImplementedError");
}

// --- tests ---
const tests = [
  { label: "example 1", input: [["aba", "cdc", "eae"]], expected: 3 },
  { label: "example 2", input: [["aaa", "aaa", "aa"]], expected: -1 },
  { label: "all length 1 different", input: [["a", "b", "c", "d"]], expected: 1 },
  { label: "all identical", input: [["abc", "abc", "abc"]], expected: -1 },
  { label: "one string longer than rest", input: [["a", "a", "ab", "abc"]], expected: 3 },
  { label: "same length different chars", input: [["abc", "abd", "abe"]], expected: 3 },
];
let passed = 0;
for (let i = 0; i < tests.length; i++) {
  const t = tests[i];
  const got = solve(t.input[0]);
  if (JSON.stringify(got) === JSON.stringify(t.expected)) {
    passed++;
    console.log(`  Test ${i + 1} (${t.label}): PASS`);
  } else {
    console.log(`  Test ${i + 1} (${t.label}): FAIL`);
    console.log(`    Expected: ${JSON.stringify(t.expected)}\n    Got:      ${JSON.stringify(got)}`);
  }
}
console.log(`\n  ${passed}/${tests.length} passed`);
process.exit(passed === tests.length ? 0 : 1);
